### ISSUE TITLE
Suggestion Backlog Slog, 1/29/2018

### ISSUE THREAD

**Issue by RyanCavanaugh** [Microsoft/TypeScript#21460]
* #20979 Classes using only private consumption of generic variables don't have correct .d.ts emit
* #21290 Literalize the flags of RegExp literals
* #20784 Support basic concat in string enum values
* #20203 `+=` should remove `undefined` from  a `string | undefined` value
* #20219 Extant properties in object literals should be treated as assignments for flow control analysis purposes
* #18523 Block-level asserts would still be nice
* #16108 Remove `undefined` from domain of `x` in `typeof x.y`
* #15550 Group errors with same text
* #21326 Allow an (abstract?) class to implement mapped types
* #20623 Allow destructuring of typed arrays
* #20703 Allow declararation (not use) of const enums in `--isolatedModules`
* #20714 Be more selective in inference between union types by attempting to pairwise match types in the source and target to avoid "cross contamination"
* #20821 Allow assertions from `U`-constrainted type parameter `T` to subtypes of `U`
* #20375 Don't produce `never` when a discriminated union from a constraint is exhausted
* #17867 Status check on "leftover" index signatures
* #16682 In `catch` block, CFA wrongly assumes assignments from `try` block *must* have never occurred

**Comment by RyanCavanaugh** [comment]
* #20979 Classes using only private consumption of generic variables don't have correct .d.ts emit
  * We need (?) to write out the type, somehow
  * But we don't actually know if the type parameter is in the property type
  * The type of the `private` property may not be speakable in the .d.ts file
    * Would be a breaking change to enforce speakability of private properties' types
  * We need to know how it's actually used to correctly determine variance
  * Only affects instances where you'd pass one `Gen<T>` to another `Gen<U>`
  * No clear fix
    * Possible hack: Emit the type parameter if it's exactly `T` (or some other type parameter)
  * Workaround is to declare a psuedoproperty
  * This bug has existed foreverish
  * Won't Fix
* #21290 Literalize the flags of RegExp literals
  * Seems like an odd corner case to care about
  * You could write a function `sticky(rgx: string) { return new RegExp(rgx, 's') }`
  * :-1:
* #20784 Support basic concat in string enum values
  * :+1:
  * Accepting PRs
* #20203 `+=` should remove `undefined` from  a `string | undefined` value
  * Add a flag?
  * Sidebar on concating `string` with non-`string`s & implicit conversions
    * Flaaaaags
  * Should we just disallow concat with `undefined` / `null` ?
    * Sidebar on implicit concat of `string` and `number`
  * `string | undefined + string | undefined` currently produces `string` but is actually `NaN` at runtime when both operands are `undefined`
  * Under `strictNullChecks` as a breaking change? With no opt-out?
  * `JSON.stringify` does return `undefined` when passed a `function` object
    * But is currently typed incorrectly as returning `string`
  * Flaaaaaaaaaaaaags
  * Ryan would rather get an error when concating `function` or `object` than `null` or `undefined`
  * Tried to force Mohamed to have an opinion but nothing happens
  * No agreement on what constitutes "better" relative to current behavior / breaking change constraints
  * Original issue is still open
* #20219 Extant properties in object literals should be treated as assignments for flow control analysis purposes
  * Discussion of implementation difficulty
     * "Doable but not trivial"
  * Restrict to bare initializations of the variable
  * Committed but no promises
  * Contingent on difficulty
* #18523 Block-level asserts would still be nice
  * The statement-form assertion has easily-understood semantics
  * Haven't found a good color for the bikeshed
  * All CFA is contingent on bind-time information
  * May or may not be bound to function calls
  * Meditate and hope for syntactic inspiration
* #16108 Remove `undefined` from domain of `x` in `typeof x.y`
  * We talked about this at a prior design meeting but no one can remember why we didn't want to do this
  * :+1:
  * Accepting PRs
* #21326 Allow an (abstract?) class to implement mapped types
  * It's really like `lateimplements` ?
  * Much easier to validate this constraint in a call position of an instance
    * Like `static implements`
  * Do these objects ever get passed anywhere? Why not validate at a call site?
  * AMF
* #20623 Allow destructuring of typed arrays
  * Why doesn't it just work already?
  * If it just works let's just go allow it
  * Assign to Ron
  * Sidebar: Complaining about Playground
* #20703 Allow declararation (not use) of const enums in `--isolatedModules`
  * Mohamed says .d.ts files shouldn't have `const enum`s
    * Yes, ban them
  * Also change logic to only error when the enum is referenced
  * :+1:
* #20821 Allow assertions from `U`-constrainted type parameter `T` to subtypes of `U`
  * This error doesn't help anyone
  * Allow cross-hierarchy comparisons for comparability relation?
  * No pure principle of reasoning to disallow certain things but not others
    * Any `T` might be a `U` because `T & U` exists
  * :+1: to adding this to the realm of plausability
  * Assign to Daniel to see if he can spec it
* #20375 Don't produce `never` when a discriminated union from a constraint is exhausted (correction: do the complete opposite of that)
  * Bug
* #16682 In `catch` block, CFA wrongly assumes assignments from `try` block *must* have never occurred
  * Use the same control flow graph for the `catch` block as for "after the try block"
  * No mechanisms in place to capture "all possible exits due to exceptions" in a control graph
  * Too Complex

* Out of time
  * #20714 Be more selective in inference between union types by attempting to pairwise match types in the source and target to avoid "cross contamination"
  * #17867 Status check on "leftover" index signatures
  * #15550 Group errors with same text
